girder/_plugin.py

```diff
@@ -0,0 +1,266 @@
+###############################################################################
+#  Copyright Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+"""
+This module defines functions for registering, loading, and querying girder plugins.
+"""
+
+import abc
+import traceback
+
+from pkg_resources import iter_entry_points
+import six
+
+from girder import logprint
+
+NAMESPACE = 'girder.plugin'
+_pluginRegistry = None
+_pluginFailureInfo = {}
+
+
+@six.add_metaclass(abc.ABCMeta)
+class GirderPlugin(object):
+    """
+    This is a base class for describing a girder plugin.  A plugin is registered by adding
+    an entrypoint under the namespace ``girder.plugin``.  This entrypoint should return a
+    class derived from this class.
+
+    Example ::
+        class Cats(GirderPlugin):
+            '''
+            This string will be displayed to the user as a plugin description.
+            '''
+            URL = 'https://girder-cats.com'
+            DEPENDENCIES = ['animals']
+
+            def load(self, info):
+                super(Cats, self).load()
+                import rest  # register new rest endpoints
+    """
+    URL = ''
+    DEPENDENCIES = []
```

Comment:
Could this be expressed as ordinary `install_requires` in `setup.py`?

Reply:
It possible, but the implementation would be somewhat complicated.  The problem is that there isn't a one-to-one relationship between python packages and girder plugins.  A plugin can depend on a package that isn't a plugin, a package can contain multiple plugins, and a plugin can depend on a package that contains a plugin without actually depending on the plugin itself.

It would be possible as a default behavior to examine the entrypoints and get the packages that contain them to generate this list automatically for the typical use case.  Plugins requiring more complicated behavior could override it.  I'm undecided if doing this is a bad idea or not.

Reply:
After in-person discussion with @jbeezley today, I expressed the opinion that:
> A plugin can depend on a package that isn't a plugin

This hopefully won't be an issue if we're able to delegate dependency tree resolution to pip, then only look at what ends up getting installed and having a girder plugin entry point (not necessarily caring how it was resolved).

> a package can contain multiple plugins

I think this should not be allowed. It needlessly complicates the situation, without offering any unique benefit (that I can think of, but I'd be happy to hear use cases).

> a plugin can depend on a package that contains a plugin without actually depending on the plugin itself

Currently, this is `staticWebDependencies`, and we only support it for plugins' web components. In the future, we could continue to support that same functionality by having a plugin depend on another in only its `package.json` (allowing it to `require` / ES6-`import` from the other), but not reference the other with a pip dependency. However, for both the Python and web worlds, it is probably best if developers reorganize their code to distribute common components in a non-Girder-requiring library package which all plugins can depend on directly.